강원대 BE_오승규 5주차 과제 (2단계)

5win opened the pull request:
안녕하세요 제이님
2단계 메시지 전송 기능 구현하고 PR드립니다!

Order 테이블을 구현하려다 하지 못했습니다.
일단 회원은 여러 주문을 가질 수 있으므로 Member : Order = 1 : N
하나의 주문에서 여러 상품을 주문할 수 있으므로 Order : Product = 1 : N 으로 구상했습니다.

회원은 여러 번 주문할 수 있고, 하나의 주문이 여러 상품을 가질 수 있기 때문에 member_id나 order_id 로는 특정 주문에 포함되는 주문 레코드들을 묶을 수 없어 별도의 주문번호와 같은 컬럼이 필요한 것 같습니다.
주문 날짜에다가 별도의 정수 생성기를 통해 2024072812345 처럼 주문번호를 생성하도록 할 예정입니다.

다음으로 옵션에 대해서 조금 고민인 부분이 있습니다.
기존에 상품 도메인이 옵션 도메인을 하위 개념으로 가지도록 만들었습니다.
그런데 주문 테이블을 만든다면 각 레코드가 어떤 상품의 어떤 옵션에 대한 주문인지에 대한 정보를 가져야 하기 때문에 Order와 Option과의 연관관계가 불가피한 것 같고 이렇게 되면 Order, Product, Option 테이블이 삼각형 처럼 연관관계를 가지게 되는데
JPA를 사용한 Order 엔티티에서도  `Long OptionId`가 아닌 `Option option`를 선언하여 객체를 저장하도록 하는게 맞을까요?

이전에 상품을 통해서 옵션에 접근하도록 의도했기에 OptionId만을 저장하고 상품 객체를 통해 OptionId에 해당하는 옵션 객체에 접근하는게 자연스러울지, Option 객체를 저장하고 해당 객체로 바로 접근하는게 자연스러울지 애매하네요..
물론 JPA를 사용하는 관점에서는 당연히 객체를 저장하는게 맞을 것 같기는 합니다.

5win committed:
feat: Add sending kakao commerce message function

5win committed:
docs: Update README.md

5win commented on src/main/java/gift/dto/KakaoCommerceMessage.java:
```diff
@@ -0,0 +1,102 @@
+    }
+}
+
+@JsonNaming(PropertyNamingStrategies.SnakeCaseStrategy.class)
```
카카오 메시지를 보내기위해 여러 객체가 필요한데 메시지 전송 내용에서만 사용하는 DTO기 때문에 이렇게 하나의 파일에 여러 클래스를 위치시키긴 했습니다.
다른 구조를 가지도록하는 것이 좋을까요?

jaymarystella replied:
넵 상관없습니다 다만 밖에서 참조가 되지않는다면 private class로 만드시는게 좋을것같아요.

5win commented on src/main/java/gift/repository/KakaoTokenRepository.java:
```diff
@@ -0,0 +1,22 @@
+import java.util.Map;
+import java.util.NoSuchElementException;
+
+public class KakaoTokenRepository {
```
카카오 API로부터 받은 토큰을 저장해놓기 위해 Map을 사용하여 메모리에 저장해두도록 했습니다.

jaymarystella replied:
잘해보셨네요! 
(여러대의 서버를 이용했을때 어떤 문제가 생길지 고민만 해보셔요)

5win replied:
세션관리 방법이 메모리에서 DB로 넘어가게된 과정과 좀 유사할 것 같습니다.
세션은 stateful하기 때문에 메모리에서 관리하다가 서버의 개수가 증가하면서 모든 서버에서 사용자의 요청을 나눠 받을 수 있도록 DB로 옮기게 됐다고 알고 있습니다.
그리고 그에 따라 추가적인 DB가 사용되는 점과 조회쿼리의 오버헤드 부분 때문에 stateless한 토큰 방식을 많이 사용한다고 알고 있습니다.

이 부분도 서버가 나눠지면 세션과 마찬가지로 DB로 옮겨서 관리해야하지 않을까 생각되네요.
추가적으로 카카오API 토큰은 민감한 정보에 속한다고 생각해서 서비스 인증 토큰에 담아서 클라이언트와 주고받도록 사용하기엔 알맞지 않다고 생각합니다.

5win commented on src/main/java/gift/service/OrderService.java:
```diff
@@ -0,0 +1,64 @@
+        );
+
+        OptionSubtractRequest optionSubtractRequest = new OptionSubtractRequest(orderRequest);
+        optionService.subtractOption(optionSubtractRequest);
```
지난주 과제에서 옵션 개수 차감을 OptionService에 만들었어서 이렇게 서비스를 주입받고 사용하도록 했습니다.
그런데 일단은 이렇게 하긴했는데 서비스에서 다른 서비스를 주입받는 것을 지양하는 것이 좋을까요?
서비스를 주입받기 보다는 OptionService에 있는 subtractOption() 메서드 기능을 OrderService로 옮겨오는 것도 괜찮을 것 같기는 합니다.

jaymarystella replied:
그러면 OrderService가 너무 비대해질것같아요. 다른 서비스 참조해도 상관없으니 지금 구조가 더 낫다고 생각합니다.

5win commented on src/main/java/gift/service/OrderService.java:
```diff
@@ -0,0 +1,64 @@
+
+        kakaoSendMessage.sendForMe(kakaoToken, new KakaoCommerceMessage(
+                product.getName(), product.getImageUrl(), orderRequest.getMessage(),
+                product.getImageUrl(), (int) product.getPrice()
```
초기 상품 가격을 Long으로 설정했는데 카카오API에서는 int형을 받도록 되어있더라구요..
일단 기능 먼저 구현하기 위해 형변환하도록 하긴 했습니다.
여러 부분을 손봐야할 것 같아서 추후에 변경하도록 하겠습니다.

jaymarystella replied:
음 넵 외부 api에 종속적인 부분이라면 어쩔 수 없는것같네요

jaymarystella commented on src/main/java/gift/domain/KakaoMessageSender.java:
```diff
@@ -0,0 +1,61 @@
+                            ErrorMessage.KAKAO_UNAUTHORIZED_MSG);
+                    case 403:
+                        throw new KakaoLoginForbiddenException(
+                            ErrorMessage.KAKAO_FORBIDDEN_MSG);
```
이것도 메서드로 빼는게 좋아보입니다

5win replied:
넵 반영했습니다!

jaymarystella commented on src/main/java/gift/dto/OrderRequest.java:
```diff
@@ -0,0 +1,40 @@
+    private Long productId;
+
+    @NotNull
+    private Long optionId;
```
질문주신 옵션 내용 관련해서,
주문정보에는 옵션 내용이 포함이 되어있어야하기 때문에 옵션엔티티를 가지는게 맞을것같습니다.

jaymarystella commented on src/main/java/gift/service/OrderService.java:
```diff
@@ -0,0 +1,64 @@
+    }
+
+    /**
+     * 메모리에 저장된 카카오 토큰, 회원, 상품의 존재 여부를 확인 후 메시지 전송. 만약 위시 리스트에 추가된 상품이라면 위시 리스트에서 삭제.
```
작성주신 순서대로 order 메서드안의 작업들을 나누어서 메서드를 각각 만들게되면 가독성이 더 좋아질 것 같습니다.

5win replied:
넵 반영하여 나누고 조회 메서드도 추가했습니다.

jaymarystella commented on src/main/java/gift/service/OrderService.java:
```diff
@@ -0,0 +1,64 @@
+    /**
+     * 메모리에 저장된 카카오 토큰, 회원, 상품의 존재 여부를 확인 후 메시지 전송. 만약 위시 리스트에 추가된 상품이라면 위시 리스트에서 삭제.
+     */
+    @Transactional
```
트랜잭션 호흡이 긴데 어떻게 개선하면 좋을지 코드 반영이 아닌 고민만 해보셔요!

5win replied:
이 부분은 좀 어려운 것 같네요..ㅎ

지금 수정한 코드상으로는 다음과 같은 로직이 한 트랜잭션 안에서 수행됩니다.
1.  회원, 상품, 옵션 존재 확인
2.  메시지 전송
3.  옵션 개수 차감
4.  주문 객체 생성 및 저장
5.  위시리스트에서 제거

여기서 반드시 한 트랜잭션 안에서 수행되어야한다고 생각되는 것은 2, 3, 4번이라고 생각합니다.
`@Transactional`이 프록시 객체랑 관련되면서 해당 클래스 객체 생성 시기랑 관련있다고 수업에서 배웠습니다.

아직 잘 이해하진 못했는데 제가 아는 선에선 이미 order 메서드가 호출된 시점에서는 동일 클래스 내의 다른 메서드에 `@Transactional`을 붙인 뒤 호출하더라도 트랜잭션이 적용되지 않는 것으로 알고 있습니다.
따라서 order 도메인 클래스와 같이 별도의 클래스에 2,3,4 로직을 넘기고 해당 클래스에서 트랜잭션을 시작하도록하면 되지 않을까 생각되네요.

보다 좋은 방법이 있다면 말씀해주시면 도움이 될 것 같습니다 :)

jaymarystella replied:
새로운 트랜잭션을 시작할 필요는 없을거같아요. 다만 메세지전송처럼 외부호출을 분리할 수 있는 방법이 있을지 고민만해보세요!

jaymarystella commented on src/main/java/gift/service/OrderService.java:
```diff
@@ -0,0 +1,64 @@
+import org.springframework.transaction.annotation.Transactional;
+
+@Service
+public class OrderService {
```
중요한 테스트가 없네요. 핵심로직을 담당하는 부분인데 테스트코드를 보충해주셔요 (다른부분들도 한번씩 확인주세요!)

5win replied:
인지하고 있었는데 아직 구현하지 못했습니다...죄송합니다
5주차 과제가 좀 늦어지는 바람에 테스트는 구현하지 못하고 리뷰드릴 것 같습니다.
6주차에서 프론트와 API 맞추고 구조를 다소 변경해야할 것 같아 마음이 좀 급하네요. 좀 늦더라도 꼭 테스트 코드는 작성해보도록 하겠습니다.

그리고 3단계 과제가 API에 대해 클라이언트와 소통하는 방법에 대한 내용인 것 같습니다.
무엇을 의도하는지 잘 모르겠어서 다른 분들이 한 내용을 보니까 스웨거를 사용하는 것 같던데 스웨거를 사용하는 방향으로 진행해보면 될까요?

jaymarystella replied:
넵 여러 방법이 있겠지만 문서화를 한다는 관점만 벗어나지 않으면 됩니다

jaymarystella reviewed (changes requested):
안녕하세요 2단계 고민 많이 해주셨습니다.
피드백 남겨드린 부분 확인 주시고 반영 부탁드릴게요!

5win committed:
refact: Change Content, Link, Commerce to private inner class

5win committed:
fix: Fix kakao message DTO and message sender
- 스네이크 케이스로 변경되지 않는 문제 해결.
- RestClient에서 toEntity()를 사용하지 않으면 잘못된 요청에 대한 오류를 발생시키지 않는 문제 해결.
	- toEntity()로 응답을 받고 로그를 출력하도록 수정함.

5win committed:
refact: Add RestClient error handler class and apply to RestClient request

5win committed:
feat: Add HTTP 5xx error handler
- KakaoLogin, KakaoMessageSender에 적용.

5win committed:
refact: Split the order method in OrderService

5win committed:
feat: Add Orders entity and add find all orders function
- 상품 주문 시 주문 테이블에 주문 정보 추가
- 모든 주문 내역 조회 기능 추가
- 과정에서 JWT 인증 부분 수정
- Orders:Member = N:1 단방향
- Orders:OrderProductOption = 1:N 양방향
- OrderProductOption:Product = N:1 단방향
- OrderProductOption:Option = N:1 단방향

5win commented on src/main/java/gift/controller/OrderController.java:
```diff
@@ -0,0 +1,44 @@
+        this.orderService = orderService;
+    }
+
+    @GetMapping("/list")
```
모든 주문 내역을 확인하는 기능을 만들었는데 인터셉터로 인해 로그인 없이는 접근이 안되는 문제가 있었습니다.
모든 주문 내역은 관리자에 한해서 보여주는게 합당해보이는데 현재 role에 대한 기능이 없어서 일단 로그인 없이 조회가 가능하도록 했습니다.
때문에 인터셉터 우회를 위해 `/list` 라는 URI를 일단 추가했습니다.

5win commented on src/main/java/gift/domain/KakaoLogin.java:
```diff
@@ -24,26 +21,16 @@ public KakaoLogin(KakaoProperties kakaoProperties) {
-                            ErrorMessage.KAKAO_FORBIDDEN_MSG);
-                }
-            })
+            .onStatus(HttpStatusCode::is4xxClientError, RestClientErrorHandler.http4xxErrorHandler)
```
KakaoLogin과 KakaoMessageSender에서 같은 로직이 중복되길래 핸들러 클래스를 따로 만들어서 4xx, 5xx 처리 핸들러를 만들고 사용하도록 수정했습니다!

5win commented on src/main/java/gift/dto/KakaoCommerceMessage.java:
```diff
@@ -0,0 +1,102 @@
+    }
+
+    @JsonNaming(PropertyNamingStrategies.SnakeCaseStrategy.class)
+    private class Content {
```
피드백 반영하여 내부 클래스로 바꾸고 private 접근 제어자를 적용했습니다!

5win committed:
chore: Remove comments

5win commented on src/main/java/gift/entity/Order.java:
```diff
@@ -0,0 +1,53 @@
+
+@Entity
+@Table(name = "orders")
+public class Order extends BaseEntity {
```
몇가지 시행착오를 거쳐 처음 PR 드린 것과는 좀 다르게 테이블 관계를 구성하게 됐습니다.
### 1. 연관관계 결정 과정
처음엔  주문과 상품이 1:N, 주문과 옵션이 1:N으로 생각했으나 다시 생각해보니 둘다 N:M이 맞는 것 같더라구요.

따라서 N:M을 구현하기 위해 중간 매핑 테이블을 만들어 OneToMany, ManyToOne으로 구현했습니다.
그런데 또 상품과 옵션에 대해 따로따로 N:M 연관관계를 만들었는데, Product와 Option 객체를 서로 다른 List로 관리하게 되기 때문에 특정 옵션이 어떤 상품에 매핑되는 것인지 알 수가 없는 문제가 있었습니다.

때문에 OrderProduct, OrderOption 두개의 중간 매핑 테이블을 만들었다가 지우고 OrderProductOption이라는 3개의 테이블 사이에서 매핑해주는 테이블을 만들어 사용했습니다.

### 2. 반정규화
현재 JPA와 연관관계를 통해 구현하면서 정규화되어 중복을 최소화하도록 테이블이 만들어졌습니다.
따라서 주문 테이블인 Orders에는 주문번호만이 존재하고, OrderProductOption 테이블을 통해 join하여 상품 정보를 얻거나 옵션 정보를 얻을 수 있습니다.
그런데 커머스 서비스의 정책에 대해서는 잘 모르나 상품, 옵션, 회원이 없어지더라도 어느정도 기간 동안은 주문 내역이 유지되어야 할 것 같다고 생각됩니다.
그런데 이렇게 정규화하여 join을 통해 주문 내역을 조회하도록 하면 PK에 해당하는 상품이 삭제되면 관련 주문 내역에서 상품 정보를 얻을 수 없을 것 같습니다.
이런 경우엔 반정규화를 통해 주문 내역은 독립적으로 유지시키는 것이 좋을까요?
예를들어 정산을 해야한다면 주문 테이블에 통 주문 금액과 같은 컬럼을 추가하든가 하는 방식이 있을 것 같습니다.

그리고 만약 반정규화를 한다라면 JPA를 통해 연관관계 매핑을 하지 못할 것 같은데 그냥 `Long productId`과 같이 엔티티 객체를 만드는 방법을 사용하면 될까요?

### 3. Orders 테이블 형태
지금은 N:M 연관관계를 위해 중간 매핑 테이블을 만들서 join하도록 구현되어있는데
처음에 생각한 방법은 Orders 테이블에 order_id, product_id, option_id 모두 가지고 있도록 구성하는 걸 생각했습니다.
특정 주문에 포함되는 상품들을 조회한다면...전부 조인하고 `where order_id = 1`과 같이 걷어내는 방법이 있긴 한데 불필요한 조인이 추가? 메모리 낭비?가 있을 것 같긴 합니다.
이 방법은 주문은 하나의 상품을 가지고, 상품은 여러 주문에 포함되기 때문에 1:N 관계인 것은 같네요.

전자의 방법이 나은 것 같다고 생각해서 구현하긴 했는데 괜찮은 선택인지 잘 모르겠습니다.

jaymarystella replied:
1. 고민을 많이 해주셨네요!
2. 실제로 row자체가 날아가는 hard delete가 아닌 soft delete형태로 삭제될것이기 때문에 문제는 없을거같아요. 이걸 구현하려면 각 엔티티에 필드하나가 더 필요하겠네요.
3.다대다 관계는 잘 푸신거같고 주문조회시 모든 정보들이 다 나와야한다면 작성주신대로 모든 정보를 포함하는것도 괜찮은것같습니다.

5win commented on src/main/java/gift/interceptor/JwtInterceptor.java:
```diff
@@ -27,19 +28,25 @@ public boolean preHandle(HttpServletRequest request, HttpServletResponse respons
         if (authorizationHeader != null && authorizationHeader.startsWith("Bearer ")) {
             token = authorizationHeader.substring(7);
+            System.out.println("token = " + token);
+        } else {
```
Authorization헤더가 null인 경우에 통과되는 문제가 있길래 이렇게 수정했습니다.

5win commented on src/main/java/gift/service/OrderService.java:
```diff
@@ -0,0 +1,114 @@
+        );
+    }
+
+    private Member findMember(String email) {
```
주입받는 객체도 좀 많은 것 같고, 여러 서비스 클래스들에서 동일한 역할의 메서드가 꽤 존재하는데요 도메인을 나눠서 로직을 서비스로부터 떼어놓는다면 중복을 제거할 수 있을 것 같다고 생각이 되네요.

이를테면 Products라는 클래스를 만들고 ProductJpaDao에 접근하는 로직을 모두 해당 클래스에 위임하고, Product 객체에 접근하는 것도 Products 도메인을 통해서만 가능하도록 변경한다면 서비스에서는 Products, Options, Members와 같은 도메인 클래스만 주입받아도 될 것 같습니다.

그러면 `products.findProductById()`와 같이 특정 메서드를 재사용도가능할 것 같고, 응집도?가 높아지는 동시에 서비스 클래스의 복잡도 낮아질 수 있을 것 같습니다. 주입받는 객체가 줄어드는 것도 한 몫할 것 같구요.

이번 주는 또 프론트와 연동을 해야하기에 이부분을 수정하긴  시간이 좀 부족할 것 같지만 나중에 개인적으로라도 구조를 수정해볼까 합니다. 아마 그때는 리뷰를 못받을 것 같아서 미리 이렇게 분리하는 방향이 어떠한지 의견 주시면 감사하겠습니다!

jaymarystella replied:
> 이를테면 Products라는 클래스를 만들고 ProductJpaDao에 접근하는 로직을 모두 해당 클래스에 위임하고, Product 객체에 접근하는 것도 Products 도메인을 통해서만 가능하도록 변경한다면 서비스에서는 Products, Options, Members와 같은 도메인 클래스만 주입받아도 될 것 같습니다.

그럼 구조가 이상해질거같습니다. 어떤 것들은 서비스를 주입받는데 어떤것들은 작성주신 ~~~s클래스를 주입받아야하니까요. 오히려 로직이 분산되어서 헷갈릴것같습니다. Product도메인의 집합을 다루는 Products클래스를 만드는것은 괜찮으나 인프라접근용으로 따로 만드는것은 개인적으로는 지양하는게 좋을것같습니다.

5win committed:
docs: Update README.md

5win committed:
feat: Add lazy fetch type and NotNull into Order, OrderProductOption entity

5win commented on src/main/java/gift/service/OrderService.java:
```diff
@@ -0,0 +1,114 @@
+    }
+
+    public Page<OrderFindAllResponse> findAllOrders(Pageable pageable) {
+        return orderProductOptionJpaDao.findAll(pageable).map(order ->
```
주문 정보 조회이기 때문에 Order 테이블로부터 조회하며 시작하는게 자연스러운 것 같긴한데 페이지네이션 때문에 order 먼저 조회하고 주문에 포함된 상품들을 가져와서 Page 객체를 구성하는 것이 다소 복잡하더라구요.

이렇게 OrderProductOption을 통해 조회하면 간단해져서 이렇게 구현했습니다.

jaymarystella reviewed (approved):
안녕하세요~
리뷰반영 고생많이 해주셨네요.
다음 단계에서 뵙겠습니다!
